### ISSUE TITLE
Enhancement: [no-misused-promises] Check when an abstract method implementation returns a promise instead of void

### ISSUE THREAD

**Issue by alythobani** [typescript-eslint/typescript-eslint#8739]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=4.8.4&showAST=es&fileType=.tsx&code=FBA&eslintrc=N4KABGBEBOCuA2BTAzpAXGYBfEWg&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
abstract class A {
  thing: number | null = null;

  abstract setThing(): void;

  setAndPrintThing(): void {
    this.setThing();
    console.log(this.thing);
  }
}

class B extends A {
  // Expected no-misused-promises error here
  async setThing(): Promise<void> {
    const thing = await fetchThing();
    this.thing = thing;
  }
}

async function fetchThing(): Promise<number> {
  return await new Promise<number>((resolve) => {
    setTimeout(() => {
      resolve(5);
    }, 1000);
  });
}

const b = new B();

// Prints null instead of 5
b.setAndPrintThing();

function _setAndPrintThing(setThing: () => void): void {
  setThing();
  console.log(b.thing);
}

// Correctly triggers no-misused-promises error:
//  "Promise returned in function argument where a void return was expected"
_setAndPrintThing(b.setThing.bind(this));
```

### ESLint Config

```javascript
{
  "rules": {
    "@typescript-eslint/no-misused-promises": "error"
  }
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

### Expected Result

Line 14 (where `B.setThing` is implemented) should show an error like "Promise returned in abstract method implementation where a void return was expected"

### Actual Result

There's no error on line 14

### Additional Info

I'm not sure why the playground link doesn't seem to be working, but hopefully you can see my code sample anyway and it makes sense.

Based on the [`no-misused-promises`](https://typescript-eslint.io/rules/no-misused-promises/) page it looks like maybe this should be another category under [`checksVoidReturn`](https://typescript-eslint.io/rules/no-misused-promises/#checksvoidreturn)?

**Comment by bradzacher** [comment]
There's no bug here - the rule doesn't inspect method return types via inheritance.

**Comment by alythobani** [comment]
Ah gotcha, sorry I didn't see the "Enhancement" option
